xmtp_cursor_state/src/store.rs

```diff
@@ -0,0 +1,227 @@
+use std::collections::HashMap;
+use std::fmt;
+use std::sync::{Arc, Mutex};
+use xmtp_proto::xmtp::xmtpv4::envelopes::Cursor;
+
+pub type SharedCursorStore = Arc<Mutex<CursorStore>>;
+
+pub struct CursorStore {
+    topics: HashMap<Vec<u8>, Cursor>,
+}
+
+impl CursorStore {
+    pub fn new() -> Self {
+        Self {
+            topics: HashMap::new(),
+        }
+    }
+
+    /// Record that a message for this topic with the given clock was processed
+    pub fn processed(&mut self, topic: Vec<u8>, new_clock: &Cursor) {
+        let current = self.topics.entry(topic).or_default();
+        current.merge_cursor(new_clock);
+    }
+
+    /// Get the current vector clock for this topic
+    pub fn get_latest(&self, topic: &[u8]) -> Option<&Cursor> {
+        self.topics.get(topic)
+    }
+
+    /// Compute the lowest common cursor across a set of topics.
+    /// For each node_id, uses the **minimum** sequence ID seen across all topics.
+    pub fn lowest_common_cursor(&self, topics: &[Vec<u8>]) -> Option<Cursor> {
+        let mut min_clock: HashMap<u32, u64> = HashMap::new();
+        let mut seen_any = false;
+
+        for topic in topics {
+            if let Some(cursor) = self.get_latest(topic) {
+                seen_any = true;
+                for (&node_id, &seq_id) in &cursor.node_id_to_sequence_id {
+                    min_clock
+                        .entry(node_id)
+                        .and_modify(|existing| *existing = (*existing).min(seq_id))
+                        .or_insert(seq_id);
+                }
+            }
+        }
+
+        if seen_any {
+            Some(Cursor {
+                node_id_to_sequence_id: min_clock,
+            })
+        } else {
+            None
+        }
+    }
+}
+
+impl fmt::Debug for CursorStore {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut entries = f.debug_map();
+
+        for (topic, cursor) in &self.topics {
+            // display topic as hex for readability
+            let topic_hex = hex::encode(topic);
+            entries.entry(&topic_hex, cursor);
+        }
+
+        entries.finish()
+    }
+}
+
+impl Default for CursorStore {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn cursor_with(kvs: &[(u32, u64)]) -> Cursor {
+        Cursor {
+            node_id_to_sequence_id: kvs.iter().cloned().collect(),
+        }
+    }
+
+    #[test]
```

Comment:
use `#[xmtp_common::test]` from the `xmtp_common` crate so that this tests runs in webassembly as well as native